src/analyzer/analyzerthread.h

```diff
@@ -0,0 +1,140 @@
+#pragma once
+
+#include <QThread>
+
+#include <atomic>
+#include <chrono>
+#include <condition_variable>
+#include <mutex>
+#include <vector>
+
+#include "analyzer/analyzerprogress.h"
+#include "analyzer/analyzer.h"
+#include "control/controlvalue.h"
+#include "preferences/usersettings.h"
+#include "sources/audiosource.h"
+#include "track/track.h"
+#include "util/db/dbconnectionpool.h"
+#include "util/samplebuffer.h"
+#include "util/memory.h"
+
+
+enum class AnalyzerThreadState {
+    Void,
+    Idle,
+    Busy,
+    Done,
+};
+
+Q_DECLARE_TYPEINFO(AnalyzerThreadState, Q_MOVABLE_TYPE);
+Q_DECLARE_METATYPE(AnalyzerThreadState);
+
+// This object lives in the creating thread of the host, i.e. does not
+// run its own event loop. It does not does not use slots for communication
+// with its host which would otherwise still be executed in the host's
+// thread.
+//
+// Signals emitted from the internal worker thread use queued connections.
+// Communication in the opposite direction is accomplished by using
+// lock-free types to avoid locking the host thread through priority
+// inversion. Lock-free types are also used for any shared state (like
+// the current analyzer progress) that is read from the host thread
+// after being notified about changes.
+//
+// The frequency of change notifications is limited to avoid flooding
+// the signal queues between the internal worker thread and the host,
+// which might cause unresponsiveness of the host.
+class AnalyzerThread : public QThread {
```

Comment:
IMHO the recommendation not to inherit from QThread only applies to the use case when running an event loop in the worker thread:

http://doc.qt.io/qt-5/qthread.html#details

"...Thus, a developer who wishes to invoke slots in the new thread must use the worker-object approach; new slots should not be implemented directly into a subclassed QThread."

AnalyzerThread has no slots to avoid queued connections and therefore does not need to run an event loop.